Capture vertex buffers

ilkkasaa opened the pull request:
Adds functionality to capture vertex buffers and the contents to a file.

ilkkasaa committed:
Capture vertex buffers

Adds functionality to capture vertex buffers and the contents to a file.

ilkkasaa committed:
Add missing changes to CMakeLists.txt

ilkkasaa committed:
Add <algorithm> for gcc

paulthomson commented on src/VkLayer_GF_amber_scoop/src/draw_call_tracker.cc:
```diff
@@ -283,4 +296,149 @@ void DrawCallTracker::CreateIndexBufferDeclarations(
+        });
+
+    // Verify that a binding description exists.
+    if (binding_description == nullptr) {
```
I think it should be `== binding_descriptions.end()`

paulthomson replied:
Also, maybe now is a good time to add a RUNTIME_ASSERT macro with an error message?

paulthomson commented on src/VkLayer_GF_amber_scoop/src/draw_call_tracker.cc:
```diff
@@ -283,4 +296,149 @@ void DrawCallTracker::CreateIndexBufferDeclarations(
+
+    // Check if the buffer has been copied already, i.e. it is stored in the
+    // |copied_buffers| map.
+    if (copied_buffers[vertex_buffer.buffer].get() == nullptr) {
```
I think this needs to be something like: `copied_buffers.count(vertex_buffer.buffer)` ? Actually, maybe `[` `]` causes an empty unique_ptr to be created? But still, it _looks_ like a mistake, and so I think it is best to avoid it, unless you make it more explicit and/or add a comment.

paulthomson commented on src/VkLayer_GF_amber_scoop/src/draw_call_tracker.cc:
```diff
@@ -283,4 +296,149 @@ void DrawCallTracker::CreateIndexBufferDeclarations(
+              device_data, vertex_buffer.buffer, buffer_create_info.size,
+              draw_call_state_.queue, GetCommandPool(device_data));
+
+      copied_buffers[vertex_buffer.buffer] = std::move(vertex_buffer_copy);
```
I guess you may as well do:

```
copied_buffers[vertex_buffer.buffer] = std::make_unique<BufferCopy>(...);
```

paulthomson commented on src/VkLayer_GF_amber_scoop/src/draw_call_tracker.cc:
```diff
@@ -283,4 +296,149 @@ void DrawCallTracker::CreateIndexBufferDeclarations(
+      copied_buffers[vertex_buffer.buffer] = std::move(vertex_buffer_copy);
+    }
+    const BufferCopy* vertex_buffer_copy =
+        copied_buffers[vertex_buffer.buffer].get();
```
Could you move this (and the above if) down to where the `vertex_buffer_copy` is actually used? Google Style suggests defining variables close to their use, in most cases.

paulthomson commented on src/VkLayer_GF_amber_scoop/src/draw_call_tracker.cc:
```diff
@@ -283,4 +296,149 @@ void DrawCallTracker::CreateIndexBufferDeclarations(
+    const BufferCopy* vertex_buffer_copy =
+        copied_buffers[vertex_buffer.buffer].get();
+
+    std::string buffer_name = "vert_";
```
The code below looks really good. But I can't help asking if we can do something simpler: I feel we should perhaps always copy the entire buffer. Then, could we also always define the buffer in AmberScript as UINT8. It looks like we then can't specify the FORMAT and OFFSET when binding the vertex data, but I think we could add those as optional parameters in Amber quite easily. I will mention this on the chat.

paulthomson commented on src/VkLayer_GF_amber_scoop/src/draw_call_tracker.cc:
```diff
@@ -283,4 +296,149 @@ void DrawCallTracker::CreateIndexBufferDeclarations(
+    std::string buffer_name = "vert_";
+    buffer_name.append(std::to_string(attribute_description.location));
+
+    VulkanFormat format = VulkanFormat(attribute_description.format);
+
+    std::string input_rate_str;
+    const VertexBufferBinding& vertex_buffer_binding =
+        draw_call_state_.bound_vertex_buffers.at(attribute_description.binding);
```
I think the above can be moved down as well.

paulthomson commented on src/VkLayer_GF_amber_scoop/include/VkLayer_GF_amber_scoop/buffer_to_file.h:
```diff
@@ -0,0 +1,57 @@
+#ifndef VKLAYER_GF_AMBER_SCOOP_BUFFER_TO_FILE_H_
+#define VKLAYER_GF_AMBER_SCOOP_BUFFER_TO_FILE_H_
+
+#include <absl/types/span.h>
```
This should be a " " include.

paulthomson commented on src/VkLayer_GF_amber_scoop/src/buffer_to_file.cc:
```diff
@@ -0,0 +1,59 @@
+
+#include "VkLayer_GF_amber_scoop/buffer_to_file.h"
+
+#include <absl/types/span.h>
```
This should be a " " include.

paulthomson reviewed (commented):
This looks good! I have reviewed most of this PR, but I have thought of a way to change the approach slightly, which may allow us to remove a lot of the code related to parsing the VkFormat.

Here is what I wrote in a comment: I can't help asking if we can do something simpler: I feel we should perhaps always copy the entire vertex buffer. Then, could we also always define the buffer in AmberScript as UINT8. It looks like we then can't specify the FORMAT and OFFSET when binding the vertex data, but I think we could add those as optional parameters in Amber quite easily. I will mention this on the chat.

ilkkasaa committed:
Use Amber script's new buffer syntax
Simplified vertex buffer copying.
Added assert macros with messages.

ilkkasaa committed:
Fix format

ilkkasaa committed:
Load vertex buffer only once

Load vertex buffer only once from the binary file and bind the same
buffer to all required locations.

paulthomson commented on src/VkLayer_GF_amber_scoop/src/vulkan_formats.cc:
```diff
@@ -0,0 +1,418 @@
+
+namespace gf_layers::amber_scoop_layer {
+
+const char* VkFormatToAmberFormatName(VkFormat vk_format) {
+  if (vk_format == VK_FORMAT_A1R5G5B5_UNORM_PACK16) {
+    return "A1R5G5B5_UNORM_PACK16";
+  }
+  if (vk_format == VK_FORMAT_A2B10G10R10_SINT_PACK32) {
+    return "A2B10G10R10_SINT_PACK32";
+  }
```
Could we use a switch here?

ilkkasaa replied:
Yeah, definitely we could. I think the reaseon I used ifs is that clang-tidy requires all enum values to be listed, not just the ones Amber supports, ie. we cannot use `default:` case. I'll check if the warning/error can be disabled with a pragma or something.

paulthomson reviewed (commented):
Looks excellent! I think we could place the ifs with a switch, but happy approve after that.

ilkkasaa committed:
Change ifs to switch-case

ilkkasaa committed:
Fix typo and disable -Wswitch-enum warning

ilkkasaa committed:
Add --http1.1 to the curl command

Try to fix the build error caused by a curl bug.

ilkkasaa committed:
Move --http1.1 curl option to proper position

paulthomson reviewed (approved):
Looks great! Thanks!
